File: tensorflow/python/compiler/tensorrt/test/topk_test.py
# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Model script to test TF-TensorRT integration."""

from tensorflow.python.compiler.tensorrt.test import tf_trt_integration_test_base as trt_test
from tensorflow.python.framework import dtypes
from tensorflow.python.framework import constant_op
from tensorflow.python.ops import nn_ops
from tensorflow.python.ops import array_ops
from tensorflow.python.platform import test


class TopKTest(trt_test.TfTrtIntegrationTestBase):
  """Testing Top-K in TF-TRT conversion."""

  def GraphFn(self, x):
    k = 5
    k_tensor = constant_op.constant(k, dtype=dtypes.int32, name="Const")
    values, indices = nn_ops.top_k(x, k_tensor, name="TopK")
    values = array_ops.identity(values, name="output_0")
    indices = array_ops.identity(indices, name="output_1")
    return values, indices

  def GetParams(self):
    k = 5
    return self.BuildParams(self.GraphFn, dtypes.float32, [[100, 100]],
                            [[100, k], [100, k]])

  def ExpectedEnginesToBuild(self, run_params):
    """Return the expected engines to build."""
    return {"TRTEngineOp_000": ["Const", "TopK"]}


class TopKOutputTypeTest(trt_test.TfTrtIntegrationTestBase):
  """Testing that output type of engine using Top-K is set correctly."""

  def GraphFn(self, x):
    k = 5
    k_tensor = constant_op.constant(k, dtype=dtypes.int32, name="Const")
    values, indices = nn_ops.top_k(x, k_tensor, name="TopK")
    # Reshape will act as a layer between the TopK output and the engine
    # output, requiring the output tensor of reshape to be set explicitly to
    # int32.
    indices = array_ops.reshape(indices, [100, 1, 5], name="Reshape")
    values = array_ops.identity(values, name="output_0")
    indices = array_ops.identity(indices, name="output_1")
    return values, indices

  def GetParams(self):
    k = 5
    return self.BuildParams(self.GraphFn, dtypes.float32, [[100, 100]],
                            [[100, k], [100, 1, k]])

  def ExpectedEnginesToBuild(self, run_params):
    """Return the expected engines to build."""
    return {"TRTEngineOp_000": ["Const", "TopK", "Reshape", "Reshape/shape"]}


if __name__ == "__main__":
  test.main()

File: tensorflow/compiler/xla/tests/topk_test.cc
/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/compiler/xla/tests/hlo_test_base.h"
#include "tensorflow/compiler/xla/tests/test_macros.h"
#include "tensorflow/compiler/xla/tests/test_utils.h"

namespace xla {
namespace {

class TopkTest : public HloTestBase {};

XLA_TEST_F(TopkTest, SimpleTopK) {
  absl::string_view hlo = R"(
HloModule topk

compare {
  p.0.lhs = bf16[] parameter(0)
  p.0.rhs = bf16[] parameter(1)
  ROOT lt = pred[] compare(p.0.lhs, p.0.rhs), direction=GT
}

ENTRY TopK {
  x = bf16[10,10] parameter(0)
  ROOT topk = (bf16[10,2], s32[10,2]) topk(x), k=2, to_apply=compare
}
)";
  EXPECT_TRUE(RunAndCompare(hlo, ErrorSpec{1e-5, 1e-5}));
}

XLA_TEST_F(TopkTest, SimpleTopKReverseDirection) {
  absl::string_view hlo = R"(
HloModule topk

compare {
  p.0.lhs = bf16[] parameter(0)
  p.0.rhs = bf16[] parameter(1)
  ROOT lt = pred[] compare(p.0.lhs, p.0.rhs), direction=LT
}

ENTRY TopK {
  x = bf16[10,10] parameter(0)
  ROOT topk = (bf16[10,2], s32[10,2]) topk(x), k=2, to_apply=compare
}
)";
  EXPECT_TRUE(RunAndCompare(hlo, ErrorSpec{1e-5, 1e-5}));
}

}  // namespace
}  // namespace xla
